### PR TITLE
[5756 Identity Management Overhaul] Placement

### PR DESCRIPTION
**Pull request by JoshVanL**
Part of https://github.com/dapr/dapr/issues/5756

Branched from https://github.com/dapr/dapr/pull/6171

Uses security package for identity and serving.

### CODE REVIEW

File: charts/dapr/charts/dapr_placement/templates/dapr_placement_statefulset.yaml [hunk 1]
```diff
@@ -177,9 +175,16 @@ spec:
                 fieldPath: metadata.namespace
       serviceAccountName: dapr-placement
       volumes:
-        - name: credentials
-          secret:
-            secretName: dapr-trust-bundle
+      - name: dapr-trust-bundle
+        configMap:
+          name: dapr-trust-bundle
+      - name: dapr-identity-token
```
Comment (ItalyPaleAle) [inline charts/dapr/charts/dapr_placement/templates/dapr_placement_statefulset.yaml:181]: What about kubelets that don't supported projected volumes? Is there a way to make this optional (so users can configure Sentry to use JWTs signed somewhere else instead)?

File: cmd/placement/options/options.go [hunk 2]
```diff
@@ -87,6 +97,10 @@ func New() *Options {
 
 	flag.Parse()
 
+	if len(*depRCF) > 0 || len(*depICF) > 0 || len(*depIKF) > 0 || len(*depCC) > 0 {
+		log.Warn("--certchain, --issuer-ca-filename, --issuer-certificate-filename and --issuer-key-filename are deprecated and will be removed in v1.12.")
```
Comment (ItalyPaleAle) [inline cmd/placement/options/options.go:101]: 1.13? Or 1.14?

File: pkg/security/security.go [hunk 3]
```diff
@@ -148,12 +160,46 @@ func (p *provider) Start(ctx context.Context) error {
...
+				log.Infof("watching trust anchors file %q for changes", p.trustAnchorsFile)
+				return fswatcher.Watch(ctx, p.trustAnchorsFile, caEvent)
+			},
+			func(ctx context.Context) error {
+				for {
+					select {
+					case <-ctx.Done():
+						return nil
+					case <-caEvent:
+						log.Info("trust anchors file changed, reloading trust anchors")
```
Comment (ItalyPaleAle) [inline pkg/security/security.go:182]: Nit:
```suggestion
						log.Info("Trust anchors file changed, reloading trust anchors")
```

File: pkg/security/security.go [hunk 4]
```diff
@@ -205,3 +259,19 @@ func SentryID(sentryTrustDomain, sentryNamespace string) (spiffeid.ID, error) {
...
+
+func (x *x509source) updateTrustAnchorFromFile(filepath string) error {
+	rootPEMs, err := os.ReadFile(filepath)
+	if err != nil {
+		return fmt.Errorf("failed to read trust anchors file %q: %s", filepath, err)
+	}
+
+	trustAnchorCerts, err := secpem.DecodePEMCertificates(rootPEMs)
+	if err != nil {
+		return fmt.Errorf("failed to decode trust anchors: %s", err)
```
Comment (ItalyPaleAle) [inline pkg/security/security.go:271]:
```suggestion
		return fmt.Errorf("failed to decode trust anchors: %w", err)
```

File: pkg/security/security.go [hunk 5]
```diff
@@ -148,12 +160,46 @@ func (p *provider) Start(ctx context.Context) error {
...
+				for {
+					select {
+					case <-ctx.Done():
+						return nil
+					case <-caEvent:
+						log.Info("trust anchors file changed, reloading trust anchors")
+
+						p.sec.source.lock.Lock()
+						if uErr := p.sec.source.updateTrustAnchorFromFile(p.trustAnchorsFile); uErr != nil {
+							log.Errorf("failed to read trust anchors file %q: %s", p.trustAnchorsFile, uErr)
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply pkg/security/security.go:186]: What's wrong with '%q'?

File: charts/dapr/charts/dapr_placement/templates/dapr_placement_statefulset.yaml (hunk 1 above)
Comment (ItalyPaleAle) [reply charts/dapr/charts/dapr_placement/templates/dapr_placement_statefulset.yaml:181]: We have at least one large customer who depends on a kubelet without support for projected volumes.

File: pkg/security/security.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pkg/security/security.go:186]: Will encode slashes so it becomes harder to read. https://go.dev/play/p/vdBrYBVp2wO

[the remark this answers is not in the document]
Comment (JoshVanL) [reply pkg/security/security.go:186]: Gotcha, makes sense. I agree :+1:

File: charts/dapr/charts/dapr_placement/templates/dapr_placement_statefulset.yaml (hunk 1 above)
Comment (JoshVanL) [reply charts/dapr/charts/dapr_placement/templates/dapr_placement_statefulset.yaml:181]: OK, can we do this in another PR as this is going to effect all the other components anyway so will be cleaner to do all of them at once in a dedicated PR?

Comment (ItalyPaleAle) [reply charts/dapr/charts/dapr_placement/templates/dapr_placement_statefulset.yaml:181]: Ok, can you please open an issue so we don't forget this is a P0?

Comment (JoshVanL) [reply charts/dapr/charts/dapr_placement/templates/dapr_placement_statefulset.yaml:181]: I've added it to the list of the parent issue https://github.com/dapr/dapr/issues/5756

**Comment by ItalyPaleAle** [comment]
Can you please test this to make sure that a Dapr 1.11 sidecar can connect to the placement service that uses the new security? We have an E2E test that tests for upgrades, but not for actors

**Comment by JoshVanL** [comment]
@ItalyPaleAle is this not the same as what the test-version-skew workflow is doing?

**Comment by JoshVanL** [comment]
> > @ItalyPaleAle is this not the same as what the test-version-skew workflow is doing?
> 
> thinking when was test-version-skew added, I didn't know about it. Yes seems like it should be ok

@ItalyPaleAle  https://github.com/dapr/dapr/pull/6374 :smile:

**Comment by JoshVanL** [comment]
Version skew daprd-sidecar-master e2e tests are failing because of a change in an error string from [this PR](https://github.com/dapr/dapr/pull/6703/files#diff-0c175408a25c2ea4a2568014b413734c677c9ab72972a6227e9b75c8b1d782e2), but otherwise this PR is passing.

**Review by artursouza — APPROVED** [review]

**Comment by MregXN** [comment]
Hi @JoshVanL and @artursouza. This PR seems to directly cause the E2E test `actor_metadata` to fail.

 Before this PR was merged, test `actor_metadata` will fail with a certain probability due to network delay. But after the PR was merged, the test has been failing.

This can be observed in [E2E test log](https://github.com/dapr/dapr/actions/workflows/dapr-test.yml) after August 15th (the date when this PR was merged) . It can be found that all `actor_metadata` tests failed with the same reason and have not succeeded since.

Could you please take a look at this issue? Thank you.

**Pull request merged**